src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -647,6 +647,12 @@ public static void Run(ServiceBase service)
 
         public void Stop()
         {
+            if (_status.currentState == ServiceControlStatus.STATE_STOPPED || _status.currentState == default)
```

Comment:
The docs for [SetServiceStatus](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-setservicestatus) mention it must never be called after the service state is stopped:
> If the status is SERVICE_STOPPED, perform all necessary cleanup and call SetServiceStatus one time only. This function makes an LRPC call to the SCM. The first call to the function in the SERVICE_STOPPED state closes the RPC context handle and any subsequent calls can cause the process to crash.

We were actually hitting this in the common case, since we'd call back into the IHostLifetime StopAsync when SCM stopped the service.  Depending on how long it was since the first stop call we were seeing crashes (this was much more common when I inserted delay).  This check just prevents that re-entrancy.

It doesn't solve a race condition where Stop might be called from multiple threads (resulting in violating this constraint) or if an application itself called Stop at same time SCM was sending the stop message.  If we wanted to solve for these cases we'd need to introduce locking.

Reply:
That latter case might actually be important -- since IHostLifetime.StopAsync gets called on a different thread and there is no guaranteed ordering that it happens after SCM's queued DeferredStop completes (that's the whole reason we had to wait in StopAsync), the two could be running concurrently.  I think we need better protection against this.